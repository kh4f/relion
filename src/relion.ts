import { bump, commit, tag, changelog } from '@/lifecycles'
import { resolveConfig, setSilent } from '@/utils'
import type { UserConfig } from '@/types'

export default async function relion(userConfig: UserConfig) {
	setSilent(!!userConfig.silent || (!!userConfig.profile && !!userConfig[`_${userConfig.profile}`]?.silent))

	const resolvedConfig = await resolveConfig(userConfig)

	const bumpResults = bump(resolvedConfig)
	const generatedChangelog = changelog(resolvedConfig)
	const commitCommand = commit(resolvedConfig)
	const tagCommand = tag(resolvedConfig)

	return { resolvedConfig, generatedChangelog, commitCommand, tagCommand, bumpResults }
}

export const defineConfig = (config: UserConfig): UserConfig => config