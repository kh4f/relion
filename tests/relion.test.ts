import { describe, expect, it } from 'vitest'
import relion from '@/.'

describe('release workflow', () => {
	it('should simulate full release workflow', () => {
		relion({
			bump: true,
			changelog: true,
			commit: true,
			tag: true,
			dryRun: true,
		})
	})
})

describe('tag generation', () => {
	it('should use default newTagFormat value', () => {
		expect(relion({ silent: true }).resolvedConfig.context.newTag).toMatch(/^v\d+\.\d+\.\d+/)
	})

	it('should use custom newTagFormat value', () => {
		expect(relion({ silent: true,
			newTagFormat: 'release-{{version}}-beta',
		}).resolvedConfig.context.newTag).toMatch(/^release-\d+\.\d+\.\d+-beta/)
	})

	it('should use custom newTagPrefix value', () => {
		expect(relion({ silent: true,
			newTagPrefix: 'relion@',
		}).resolvedConfig.context.newTag).toMatch(/^relion@\d+\.\d+\.\d+/)
	})

	it('should prefer newTagPrefix over newTagFormat', () => {
		expect(relion({ silent: true,
			newTagPrefix: 'relion@',
			newTagFormat: 'release-{{version}}-beta',
		}).resolvedConfig.context.newTag).toMatch(/^relion@\d+\.\d+\.\d+/)
	})

	it('should use custom newTagPrefix value', () => {
		expect(relion({ silent: true,
			newTagPrefix: 'relion@',
		}).resolvedConfig.context.newTag).toMatch(/^relion@\d+\.\d+\.\d+/)
	})

	it('should prefer newTagPrefix over newTagFormat', () => {
		expect(relion({ silent: true,
			newTagPrefix: 'relion@',
			newTagFormat: 'release-{{version}}-beta',
		}).resolvedConfig.context.newTag).toMatch(/^relion@\d+\.\d+\.\d+/)
	})
})