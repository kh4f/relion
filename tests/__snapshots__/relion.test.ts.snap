// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`version bumping > should append new version entry to 'versions.json' 1`] = `
"{
	"1.10.1": "0.15.0",
	"1.10.2": "0.15.0",
	"1.11.0": "0.15.0",
	"1.11.1": "0.15.0",
	"1.11.2": "0.15.0",
	"1.2.3": "0.15.0"
}"
`;
